test-data/unit/check-based-intersection.test

```diff
@@ -0,0 +1,340 @@
+class A: pass
+class B: pass
+class C: pass
+
+x: A
+reveal_type(x)  # N: Revealed type is "__main__.A"
+assert isinstance(x, B)
+reveal_type(x)  # N: Revealed type is "__main__.A & __main__.B"
+assert isinstance(x, C)
+reveal_type(x)  # N: Revealed type is "__main__.A & __main__.B & __main__.C"
+
+y: (A & B) | (A & C)
+assert isinstance(y, C)
+reveal_type(y)  # N: Revealed type is "__main__.A & __main__.B & __main__.C | __main__.A & __main__.C"
+[builtins fixtures/tuple.pyi]
+
+
+-- This is not implemented yet, maybe next sprint
+[case testInvalidIntersection-xfail]
+from __future__ import annotations
+
+a: int & str  # E: Intersection of "int & str" cannot exist: would have incompatible method signatures  [unreachable]
+a: 1 & 2  # E: Intersection of "1 & 2" cannot exist: would have incompatible method signatures  [unreachable]
+
+
+[case testIntersectionWithAny]
+# flags: --allow-any-expr --allow-any-explicit --disallow-redefinition
+from __future__ import annotations
+
+from typing import Any
+
+a: Any & str
+reveal_type(a)  # N: Revealed type is "Any & str"
+a = 1  # E: Incompatible types in assignment (expression has type "int", variable has type "Any & str")  [assignment]
+a = ""
+
+b: Any
+assert isinstance(b, str)
+reveal_type(b)  # N: Revealed type is "str"
+[builtins fixtures/tuple.pyi]
+
+
+[case testOverrideInIntersection]
+from __future__ import annotations
+
+class A:
+    def foo(self) -> int:
+        return True
+class B(A):
+    def foo(self) -> bool:
+        return True
+
+b: A & B
+reveal_type(b.foo())  # N: Revealed type is "bool"
+[builtins fixtures/tuple.pyi]
+
+
+[case testOverloaded]
+from __future__ import annotations
+from typing import overload
+class A:
+    @overload
+    def foo(self): ...
+    @overload
+    def foo(self, x: int): ...
+    def foo(self, x=1): ...
+class B:
+    def foo(self, x: str): ...
+
+x: A & B
+x.foo(None)  # E: No overload variant of "foo" of "__main__.A & __main__.B" matches argument type "None"  [call-overload] \
+             # N: Possible overload variants: \
+             # N:     def foo(self) -> None \
+             # N:     def foo(self, x: int) -> None \
+             # N:     def foo(self, x: str) -> None
+[builtins fixtures/tuple.pyi]
+
+
+[case testGenericFunction]
+from __future__ import annotations
+
+from typing import TypeVar
+
+T = TypeVar('T')
+U = TypeVar('U')
+
+def foo(x: T, y: U) -> T & U: pass
+a: int
+b: object
+reveal_type(foo(a, b))  # N: Revealed type is "int"
+[builtins fixtures/tuple.pyi]
+
+
+[case testIntersectionInferenceWithTypeVarValues]
+from __future__ import annotations
+from typing import TypeVar
+AnyStr = TypeVar('AnyStr', bytes, str)
+class A: pass
+strA: str & A
+def f(x: AnyStr & A, *a: AnyStr): pass
+f(strA)
+f(strA, 'bar')
+f(strA, b'bar') # E: Value of type variable "AnyStr" of "f" cannot be "str | bytes"  [type-var]
+[builtins fixtures/primitives.pyi]
+[builtins fixtures/tuple.pyi]
+
+
+[case testIntersectionTwoPassInference-skip]
```

Comment:
does `-skip` mean skip the test? should it be xfail instead?

Reply:
lol, I think that's what I originally intended, but it works, so.